src/target/riscv/riscv.c

```diff
@@ -4945,8 +4945,11 @@ int riscv_get_register(struct target *target, riscv_reg_t *value,
  */
 int riscv_save_register(struct target *target, enum gdb_regno regid)
 {
-	assert(target->state == TARGET_HALTED &&
-			"Doesn't make sense to populate register cache on non-halted targets.");
+	if (target->state != TARGET_HALTED) {
+		LOG_TARGET_ERROR(target, "Can't save register %s on a hart that is not halted.",
+				 gdb_regno_name(regid));
```

Comment:
If you like, the state of the target can be part of the error message. This can help in future with log analysis troubleshooting.

```suggestion
		LOG_TARGET_ERROR(target, "Can't save register %s on a hart that is not halted (current state: %s).",
				 gdb_regno_name(regid), target_state_name(target));
```